include/core/lf_token.h

```diff
@@ -0,0 +1,362 @@
+ * to be interpreted as an array of the given length. It has a pointer to type
+ * (token_type_t), which has an element_size field specifying the size of each
+ * array element (or the size of the whole value if it is not an array and has
+ * length 1). The type also optionally has function pointers to a destructor
+ * and copy constructor. These must be specified if the payload (value) is a complex
+ * struct that cannot be freed by a simple call to free() or copied by a call
+ * to memcpy().
+ * 
+ * An instance of a port struct and trigger_t struct (an action or an input port)
+ * can be cast to token_template_t, which has a token_type_t field called type
+ * and a pointer to a token (which may be NULL).  The same instance can also be
+ * cast to token_type_t, which has an element_size field and (possibly) function
+ * pointers to a destructor and a copy constructor.
+ * 
+ * A "template token" is one pointed to by a token_template_t (an action or a port).
+ * Such a token will persist in the template until it is overwritten by another
+ * token. Every token_template_t gets initialized with such a token.
+ * Before that token is used the first time, its reference count will be 0.
+ * Once it has been assigned a value, its reference count will be 1.
+ * When the token_template_t (port or action) is assigned a new value, if
+ * the reference count is 1, then the same token will be reused, and
+ * any previous value (payload) will be freed.
+ */
+
+#ifndef LF_TOKEN_H
+#define LF_TOKEN_H
+
+#include <stdlib.h> // Defines size_t
+
+//////////////////////////////////////////////////////////
+//// Constants and enums
+
+/** Possible return values for _lf_done_using and _lf_free_token. */
+typedef enum token_freed {
+    NOT_FREED = 0, // Nothing was freed.
+    VALUE_FREED,   // The value (payload) was freed.
+    TOKEN_FREED,    // The token was freed but not the value.
+    TOKEN_AND_VALUE_FREED // Both were freed
+} token_freed;
+
+//////////////////////////////////////////////////////////
+//// Data structures
+
+/**
+ * @brief Type information for tokens.
+ * Specifically, this struct contains the fields needed to support
+ * token types, which carry dynamically allocated data.
+ */
+typedef struct token_type_t {
+    /** Size of the struct or array element. */
+    size_t element_size;
+    /** The destructor or NULL to use the default free(). */
+    void (*destructor) (void* value);
+    /** The copy constructor or NULL to use memcpy. */
+    void* (*copy_constructor) (void* value);
+} token_type_t;
+
+/**
+ * Token type for dynamically allocated arrays and structs sent as messages.
+ *
+ * This struct is the wrapper around the dynamically allocated memory
+ * that carries the message.  The message can be an array of values,
+ * where the size of each value is element_size (in bytes). If it is
+ * not an array, or is not to be treated as an array, the length == 1.
+ *
+ * In the C LF target, a type for an output that ends in '*' or '[]' is
```

Comment:
It would be great if we implement @oowekyala's suggested syntax for token types in LF because this, I feel, is too hacky.